Introduce WTF::flatMap

aestes opened the pull request:
#### f3d2011c12f9cfae2515edca040e24ef0fcd37a3
<pre>
Introduce WTF::flatMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=261780">https://bugs.webkit.org/show_bug.cgi?id=261780</a>
rdar://problem/115751217

Reviewed by Richard Robinson.

Introduced WTF::flatMap. Similar to map and compactMap, flatMap takes a collection that may contain
nested collections and allows the caller to specify a map function that produces a flat Vector.

Added test cases to TestWTF. A follow-on change will adopt this in CDMInstanceFairPlayStreamingAVFObjC.

* Source/WTF/wtf/Vector.h:
(WTF::FlatMapper::flatMap):
(WTF::flatMap):
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
(TestWebKitAPI::mapVector):
(TestWebKitAPI::TEST):
(TestWebKitAPI::mapInnerStruct):

Canonical link: <a href="https://commits.webkit.org/268171@main">https://commits.webkit.org/268171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/441954701cb84281f94646d104598874479dbbb7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18893 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19840 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20762 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17672 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19084 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22535 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19366 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19118 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19240 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/16456 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/21675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16455 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17221 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23640 "layout-tests (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/16393 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17489 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17393 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21568 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/18217 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17975 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/22270 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17049 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5417 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4492 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21412 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23516 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17811 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5276 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1898,6 +1898,42 @@ Vector<typename CompactMapper<MapFunction, SourceType>::DestinationItemType> com
+};
+
+template<typename MapFunction, typename SourceType>
+struct FlatMapper<MapFunction, SourceType, typename std::enable_if<std::is_rvalue_reference<SourceType&&>::value>::type> {
```
can this use concepts instead of SFINAE?

aestes replied:
TBH I'm not familiar with concepts, but maybe! I just copied this implementation from `CompactMapper`.

rr-codes replied:
I'll try to help out!

aprotyas reviewed (approved):
LGTM, and this would be a well-contained piece of code to introduce concepts to. @rr-codes what do you think about landing this as-is and then you can do the SFINAE->Concepts transformation? That unblocks @aestes from adopting this into CDMInstanceFairPlayStreamingAVFObjC

rr-codes commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1898,6 +1898,42 @@ Vector<typename CompactMapper<MapFunction, SourceType>::DestinationItemType> com
+};
+
+template<typename MapFunction, typename SourceType>
+struct FlatMapper<MapFunction, SourceType, typename std::enable_if<std::is_rvalue_reference<SourceType&&>::value>::type> {
```
```suggestion
template<typename MapFunction, typename SourceType> requires std::is_rvalue_reference<SourceType&&>::value
struct FlatMapper<MapFunction, SourceType> {
```

rr-codes commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1898,6 +1898,42 @@ Vector<typename CompactMapper<MapFunction, SourceType>::DestinationItemType> com
     return CompactMapper<MapFunction, SourceType>::compactMap(std::forward<SourceType>(source), std::forward<MapFunction>(mapFunction));
 }
 
+template<typename MapFunction, typename SourceType, typename Enable = void>
```
```suggestion
template<typename MapFunction, typename SourceType>
```

(with the concept)

aestes committed:
Introduce WTF::flatMap
https://bugs.webkit.org/show_bug.cgi?id=261780
rdar://problem/115751217

Reviewed by Richard Robinson.

Introduced WTF::flatMap. Similar to map and compactMap, flatMap takes a collection that may contain
nested collections and allows the caller to specify a map function that produces a flat Vector.

Added test cases to TestWTF. A follow-on change will adopt this in CDMInstanceFairPlayStreamingAVFObjC.

* Source/WTF/wtf/Vector.h:
(WTF::FlatMapper::flatMap):
(WTF::flatMap):
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
(TestWebKitAPI::mapVector):
(TestWebKitAPI::TEST):
(TestWebKitAPI::mapInnerStruct):

Canonical link: https://commits.webkit.org/268171@main
